From: Rene Kita <mail@rkta.de>
Date: Sun Sep 22 13:29:27 CEST 2024
Subject: Allow unknown tags to have an id
Patch: patches/Allow-unknown-tags-to-have-an-id.patch

---
 html.c      |    2 +-
 parsetagx.c |   10 +++++++---
 2 files changed, 8 insertions(+), 4 deletions(-)

--- a/html.c
+++ b/html.c
@@ -88,7 +88,7 @@ unsigned char ALST_IMG_ALT[] =
 unsigned char ALST_NOP[] = { ATTR_CORE };
 
 TagInfo TagMAP[MAX_HTMLTAG] = {
-    {NULL, NULL, 0, 0},						/*   0 HTML_UNKNOWN    */
+    {NULL, ALST_ID, ARR_SZ(ALST_ID), 0},						/*   0 HTML_UNKNOWN    */
     {"a", ALST_A, ARR_SZ(ALST_A), 0},				/*   1 HTML_A          */
     {"h", ALST_P, ARR_SZ(ALST_P), 0},				/*   2 HTML_H          */
     {"p", ALST_P, ARR_SZ(ALST_P), 0},				/*   3 HTML_P          */
--- a/parsetagx.c
+++ b/parsetagx.c
@@ -35,7 +35,8 @@ static int (*toValFunc[]) (char *, void
 static int
 noConv(char *oval, void *str)
 {
-    *(char **)str = oval;
+    if (str)
+	*(char **)str = oval;
     return 1;
 }
 
@@ -140,8 +141,7 @@ parse_tag(char **s, int internal)
 
     tag_id = getHash_si(&tagtable, tagname, HTML_UNKNOWN);
 
-    if (tag_id == HTML_UNKNOWN ||
-	(!internal && TagMAP[tag_id].flag & TFLG_INT))
+    if (!internal && TagMAP[tag_id].flag & TFLG_INT)
 	goto skip_parse_tagarg;
 
     tag = New(struct parsed_tag);
@@ -272,6 +272,10 @@ parse_tag(char **s, int internal)
     if (*q == '>')
 	q++;
     *s = q;
+    if (!tag
+	|| (tag->tagid == HTML_UNKNOWN
+	    && !parsedtag_get_value(tag, ATTR_ID, NULL)))
+	return NULL;
     return tag;
 }
 
